### PR TITLE
fsdp with custom process groups

### PR DESCRIPTION
**Pull request by vchiley**
# What does this PR do?

This PR enables FSDP _auto_wrap to accept and propagate custom arguments, most notably, it enables the user to propagate dist process_groups.
Custom process_groups enable users to instantiate fsdp modules which shard / synchronize parameters over a subset of accelerators. This is useful for MoE Expert layers and TensorParallel layers.

# What issue(s) does this change relate to?

Mixture of Experts (MoE)
https://mosaicml.atlassian.net/browse/RESEARCH-351
https://mosaicml.atlassian.net/browse/CO-1716
https://mosaicml.atlassian.net/browse/CO-1715
https://mosaicml.atlassian.net/browse/CO-1714
https://mosaicml.atlassian.net/browse/CO-1712

TensorParallel (TP)
https://mosaicml.atlassian.net/browse/CO-1635
https://mosaicml.atlassian.net/browse/RESEARCH-442

In general this doesn't fix an issue, but enables a feature which enable TP and MoE models
https://github.com/mosaicml/examples/pull/180
also see: https://github.com/mosaicml/tutel/pull/1

btw I ran [this](https://github.com/mosaicml/examples/pull/180/files#diff-6db87bde9cfb73791e1be455afa625ad3dabc4ea723aa1eb56a10c5b83240914R213) "test" on 16 gpus with different process_group configurations to validate different configurations run. For the MoE setup with pg 'self' or 'set1', I do test if the different configurations result in the expert weights not being sync'd. I do not check if params / grads are sync'd in appropriately in the TP setting; since TP params are sharded, the params shouldn't be equal, so idk what to check, just the fact that they params are gathered, have the appropriate shape, and run might be test enough.

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Did you update any related docs and document your change?
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

Tests: this PR enables "super-users" to use advanced features and defaults to previous behavior. All former tests passing is test enough.

### CODE REVIEW

**Comment by dskhudia** [comment]
This also allows user to specify per module sharding strategy as well?

File: composer/trainer/mosaic_fsdp.py
Comment (dskhudia) [inline composer/trainer/mosaic_fsdp.py:86]: It's immediately not clear the pro is shorted version of process. Consider changing name to custom_recursive_wrap?

File: composer/trainer/mosaic_fsdp.py
Comment (dskhudia) [inline composer/trainer/mosaic_fsdp.py:109]: Missing args process_group_cache and others here.

**Comment by vchiley** [comment]
https://github.com/mosaicml/composer/pull/2006#issuecomment-1445743125
Yes this enables the user to pass custom args to fsdp for every fsdp wrapped module; we include some of the mosaicml tooling for easily setting the configs
cc @dskhudia

File: composer/trainer/mosaic_fsdp.py
Comment (bcui19) [inline composer/trainer/mosaic_fsdp.py:58]: Do we want to throw a warning here for now? + a TODO when fp8 is supported in PyTorch?

Comment (vchiley) [reply composer/trainer/mosaic_fsdp.py:58]: This was [copy pasta from dist_strategy.py](https://github.com/mosaicml/composer/pull/2006/files/1e9ebe27db16f25c3528dbd8d4c53c6652f49757#diff-dd51f7e9f15c34f3f9e4895d122b92bf4e8e713cdf582a4af5e5e2eef628c0d1L126), but yes adding a warning sounds like a good idea. will do.

File: composer/trainer/mosaic_fsdp.py
Comment (bcui19) [inline composer/trainer/mosaic_fsdp.py:138]: Don't we need an equality check `== 0` here?

File: composer/trainer/mosaic_fsdp.py
Comment (bcui19) [inline composer/trainer/mosaic_fsdp.py:145]: same as above with set?

File: composer/trainer/mosaic_fsdp.py
Comment (vchiley) [reply composer/trainer/mosaic_fsdp.py:141]: if `world_size` is divisible by `k` then `world_size % k == 0` and we do not need to raise the error.
with an `== 0` check you'd also need to add a not
```
if not world_size % k == 0:
```

right?

**Comment by bcui19** [comment]
LGTM, will approve when we add docs

**Review by mvpatel2000 — CHANGES_REQUESTED** [review]
I'm going to hold this PR until we get multiple approval here just because it looks scary and might break release, so we should be extra careful. Will lift once we have sign-offs from all parties. This is basically in lieu of codeowners for this part of the repo, which isn't configured.

Requiring approvals from @bcui19 @dakinggg and myself.

File: composer/trainer/mosaic_fsdp.py
Comment (dakinggg) [reply composer/trainer/mosaic_fsdp.py:141]: `world_size % k != 0` would be a bit clearer I think, just so I don't have to think about which case is which

File: composer/trainer/mosaic_fsdp.py
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:159]: I don't understand the second sentence here. can you clarify?

File: composer/trainer/mosaic_fsdp.py
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:282]:
```suggestion
        """Updates _auto_wrap to enable module_kwargs.
```

File: composer/trainer/mosaic_fsdp.py
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:186]: could you call out what specifically is modified?

File: composer/trainer/mosaic_fsdp.py
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:284]: could you call out what specifically is modified?

File: composer/trainer/mosaic_fsdp.py
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:29]: these aren't really private functions since they're being exported, so maybe remove the `_` and add docstrings?

**Review by dakinggg — COMMENTED** [review]
I think this looks good to me, left a few minor comments. Could you also test that a mosaicgpt-125m runs and loss curve looks reasonable? (i.e. make sure this PR doesn't break stuff unrelated to this PR)

**Comment by vchiley** [comment]
@bcui19 added docs
@dakinggg made suggested updts, suggested runs from diff branches [here](https://wandb.ai/mosaic-ml/composer_mfsdp_test) (using [tutel_moe branch](https://github.com/vchiley/examples/tree/tutel_moe) in examples repo since it points to this branch)

File: docs/source/notes/distributed_training.rst
Comment (dakinggg) [inline docs/source/notes/distributed_training.rst:309]:
```suggestion
**Experimental:** Composer enables users to specify custom FSDP args for all wrapped modules. This is enabled by returning a dictionary of args instead of returning a bool.
```

File: docs/source/notes/distributed_training.rst
Comment (dakinggg) [inline docs/source/notes/distributed_training.rst:324]: maybe this model should actually contain some `BlockRequiringCustomArgs`?

File: docs/source/notes/distributed_training.rst
Comment (dakinggg) [inline docs/source/notes/distributed_training.rst:355]:
```suggestion
2. :code:`'node'`: instantiates process groups which opereate within a node (:code:`'node'` == :code:`f'set{local_world_size}'`). This is useful for Expert Layers in MoE models.
```

File: composer/trainer/mosaic_fsdp.py
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:281]:
```suggestion
    """Updates FSDP's _auto_wrap to enable module_kwargs."""
```

**Review by dakinggg — APPROVED** [review]
LGTM

File: docs/source/notes/distributed_training.rst
Comment (vchiley) [reply docs/source/notes/distributed_training.rst:324]: valid point

**Review by bcui19 — APPROVED** [review]
LGTM!

**Review by mvpatel2000 — APPROVED** [review]
LGTM.

**Pull request merged**
